### DISCUSSION TITLE
Help with Builder instancing

### DISCUSSION THREAD

**Discussion by lufriem** [vsg-dev/VulkanSceneGraph#935] [Q&A] [answered]
Hi, please forgive my ignorance, I've been looking at the vsgBuilder example, with instancing enabled, like originally discussed [in the vsg-users google group](https://groups.google.com/g/vsg-users/c/ENqWY1J2rug/m/A1oXe3CfAQAJ). There was [an additional discussion on how to manipulate (move) these instances](https://groups.google.com/g/vsg-users/c/-gzdrv9gVU0/m/oWyREVPWAQAJ).

So what I tried to achieve was to simply change the color and position of sphere instances. Rather then creating 10000 I thought I'd start small, to help debugging, so I settled for three spheres

> vsgbuilder --sphere -n 3

Now, in the second vsg-users thread, @robertosfield mentioned

> To update the positions after vsg::Builder has set things up will require you to traverse the subgraph to find the VertexIndexDraw node then travers it's arrays to find the vsg::BufferInfo object that has data equal to the GeoetryInfio.positions vec3array.  This BufferInfo is the one you'll need to use when updating the array using vsg::CopyAndReleaseBuffer.

So what I did was (naively, I guess)
1. set up a new update operation to my viewer (vsg::Viewer::addUpdateOperation with the vsg::UpdateOperations::ALL_FRAMES parameter)
2. which in turn makes the scene apply a visitor traversing all nodes
3. In this visitor I check for the VertexIndexDraw node (there's only one in my simple example! I can confirm its instance count of 3) and inspect its arrays, from the Builder::createSphere code I could see that the color data is in the BufferInfoList array index 3 (position would be index 4).
4. So far so good, using the debugger I could confirm that the BufferInfo's Data indeed represents the colors/positions set when the scene was built.
5. But this is where I am stuck, since vsg::CopyAndReleaseBuffer seems to have been deprecated

> /// Deprecated - use vsg::Data dataVariance and Data::dirty() to signal vsg::TransferTask to transfer data.

So my question is, **what exactly do I have to do now?**
- set the dataVariance in the BufferInfo's Data properties member to vsg::DataVariance::DYNAMIC_DATA? Is it 'good enough' to set the dataVariance in the visitor, or does this need to be set when initially creating the instances (I'd guess the Builder class would have done that though..)
- and call the 'dirty()' function on the BufferInfo in the visitor, once I'm done with my changes?
- How do I actually set my new values (different colors, or different positions)? Can I simply overwrite whatever's in the Data buffer?

Right now none of my changes seem to result in any on-screen difference (even though the Data is changed between Visitor calls..), I'm sure I'm doing something stupid somewhere..

Any help/pointers would be appreciated!

**Comment by robertosfield** [comment, marked as answer]
Setting  vsg::Data::properties.dataVariance to DYNAMIC_DATA and then
calling data->dirty() is the way to tell the VSG that data is updated and
needs transferring to the associated GPU memory.

This transfer is done by the vsg::TransferData which is a container that
the vsg::Viewer will set up internally for during the compilation of scene
graph that happens when viewer->compile() is first called.  This means you
should set up the Data::properties.dataVariance before the
viewer->compile() call for things to be set up correctly.

The vsgdynamcvertex and vsgdynamictexture and vsgdynamictexture_cs and
vsgclip are all examples that illustrate how to manage dynamic data.

The old references to vsg::CopyAndReleaseBuffer are best forgotten as they
are more complex to use and less efficient that the newer vsg::TransferData
class that tracks the dynamic data's modifiedCount() and transfer data when
data has been modified  - singated by calling data->dirty().

It's a while since I worked on vsgbuilder so can't comment on the specifics
of what might be supported, but I think the instance container should be
shared by the built scene graph so it should be possible to set up the
positions data.

**Reply by lufriem** [reply]
Thank you for the prompt reply! Indeed, setting the dataVariance on the buffer **before** calling compile was the missing piece! I can now change both position and color of specific instances..
